pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,842 @@
+		content.Spec.VolumeSnapshotRef.UID == snapshot.UID {
+		return true
+	}
+	return false
+}
+
+// The function checks whether the volumeSnapshotRef in snapshot content matches the given snapshot. If match, it binds the content with the snapshot
+func (ctrl *csiSnapshotController) checkandBindSnapshotContent(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) error {
+	if content.Spec.VolumeSnapshotRef == nil || content.Spec.VolumeSnapshotRef.Name != snapshot.Name {
+		return fmt.Errorf("Could not bind snapshot %s and content %s, the VolumeSnapshotRef does not match", snapshot.Name, content.Name)
+	} else if content.Spec.VolumeSnapshotRef.UID != "" && content.Spec.VolumeSnapshotRef.UID != snapshot.UID {
+		return fmt.Errorf("Could not bind snapshot %s and content %s, the VolumeSnapshotRef does not match", snapshot.Name, content.Name)
+	} else if content.Spec.VolumeSnapshotRef.UID == "" {
+		contentClone := content.DeepCopy()
+		contentClone.Spec.VolumeSnapshotRef.UID = snapshot.UID
+		contentClone.Spec.VolumeSnapshotClassName = snapshot.Spec.VolumeSnapshotClassName
+		newContent, err := ctrl.clientset.VolumesnapshotV1alpha1().VolumeSnapshotContents().Update(contentClone)
+		if err != nil {
+			glog.V(4).Infof("updating VolumeSnapshotContent[%s] error status failed %v", newContent.Name, err)
+			return err
+		}
+		_, err = ctrl.storeContentUpdate(newContent)
+		if err != nil {
+			glog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", newContent.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (ctrl *csiSnapshotController) checkandUpdateSnapshotStatusOperation(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshot, error) {
+	status, _, err := ctrl.handler.GetSnapshotStatus(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check snapshot status %s with error %v", snapshot.Name, err)
+	}
+
+	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, status, time.Now(), nil, IsSnapshotBound(snapshot, content))
+	if err != nil {
+		return nil, err
+	}
+	return newSnapshot, nil
+}
+
+// The function goes through the whole snapshot creation process.
+// 1. Trigger the snapshot through csi storage provider.
+// 2. Update VolumeSnapshot status with creationtimestamp information
+// 3. Create the VolumeSnapshotContent object with the snapshot id information.
+// 4. Bind the VolumeSnapshot and VolumeSnapshotContent object
+func (ctrl *csiSnapshotController) createSnapshotOperation(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshot, error) {
+	glog.Infof("createSnapshot: Creating snapshot %s through the plugin ...", snapshotKey(snapshot))
+
+	if snapshot.Status.Error != nil && !isControllerUpdateFailError(snapshot.Status.Error) {
+		glog.V(4).Infof("error is already set in snapshot, do not retry to create: %s", snapshot.Status.Error.Message)
+		return snapshot, nil
+	}
+	class, err := ctrl.GetClassFromVolumeSnapshot(snapshot)
```

Comment:
We may update snapshot when the snapshot  do not set VolumeSnapshotClassName field and use the default  snapshot class , in this case, the follow update will always be version conflict, so we`d better return the snapshot from the function.